daemon/cmd/daemon.go

```diff
@@ -274,7 +275,61 @@ func restoreCiliumHostIPs(ipv6 bool, fromK8s net.IP) {
 		fromFS = node.GetInternalIPv4Router()
 	}
 
-	node.RestoreHostIPs(ipv6, fromK8s, fromFS, cidr)
+	restoredIP := node.RestoreHostIPs(ipv6, fromK8s, fromFS, cidr)
+	if err := removeOldRouterState(restoredIP); err != nil {
+		log.WithError(err).Warnf(
+			"Failed to remove old router IPs (restored IP: %s) from cilium_host. Manual intervention is required to remove all other old IPs.",
+			restoredIP,
+		)
+	}
+}
+
+// removeOldRouterState will try to ensure that the only IP assigned to the
+// `cilium_host` interface is the given restored IP. If the given IP is nil,
+// then it attempts to clear all IPs from the interface.
+func removeOldRouterState(restoredIP net.IP) error {
+	l, err := netlink.LinkByName(defaults.HostDevice)
+	if err != nil {
+		return err
+	}
+
+	family := netlink.FAMILY_V6
+	if restoredIP.To4() != nil {
+		family = netlink.FAMILY_V4
+	}
+	addrs, err := netlink.AddrList(l, family)
+	if err != nil {
+		return err
+	}
+	addrLen := len(addrs)
+	switch {
+	case addrLen == 1 && restoredIP == nil:
+		if err := netlink.AddrDel(l, &addrs[0]); err != nil {
+			return fmt.Errorf("failed to remove only router IP: %w", err)
+		}
+	case addrLen > 1:
+		log.Info("More than one router IP was found on the cilium_host device after restoration, cleaning up old router IPs.")
+
+		var errs []error
+		for _, a := range addrs {
+			if restoredIP == nil {
+				goto deleteIP
+			}
+			if restoredIP.Equal(a.IP) {
+				continue
+			}
```

Comment:
I'm probably missing something, but why we cannot simplify the logic like the below?
```
	addrs, err := netlink.AddrList(l, family)
	if err != nil {
		return err
	}

	for _, a := range addrs {
		if  restoredIP != nil && restoredIP.Equal(a.IP) {
			continue
		}
		if err := netlink.AddrDel(l, &a); err != nil {
				errs = append(errs, fmt.Errorf("failed to remove IP %s: %w", a.IP, err))
		}
	}
```

Reply:
Thanks Daniel, good suggestion. I didn't notice it can be simplified. [Diff](https://github.com/cilium/cilium/compare/6844f420b23b0c51c4b67bbc791f9bfeedb38dc2..cd25bc1153a746903ff2e5f50f3c466cfd6216aa) pushed.